src/model_server/lib/unity_global.cpp

```diff
@@ -477,6 +477,19 @@ namespace turi {
     }
   }
 
+  toolkit_function_response_future unity_global::run_toolkit_background(
+      std::string toolkit_name,
+      variant_map_type& _arguments) {
+
+    auto capture_args = std::make_shared<variant_map_type>(_arguments);
```

Comment:
why this copy is needed?

Reply:
Because the arguments are used in another thread after the call below returns, and it's passed by reference into this function, it may be invalid when used later.  Thus the function needs to capture all arguments.

Reply:
Yeah, totally understand that. I mean, since you are using `[=]` to capture by value, can't `_arguments` be copied implicitly?

```cpp
return toolkit_function_response_future(
         [=]() {
            return this->run_toolkit(toolkit_name, _arguments);
         });
   }
```

Reply:
So I learned a minor detail of the C++ standard just now governing this.  I assumed the reference was the type used in the value capture, but it turns out when that type is a reference it's the referenced type that's used.  So I don't need that statement: 
```
5.1.2 Lambda expressions

(...) 14. An entity is captured by copy if it is implicitly captured and the capture-default is = or if it is explicitly captured with a capture that does not include an &. For each entity captured by copy, an unnamed nonstatic data member is declared in the closure type. The declaration order of these members is unspecified. The type of such a data member is the type of the corresponding captured entity if the entity is not a reference to an object, or the referenced type otherwise. 
```

Reply:
glad to know that! But this is still not quite safe because the value is referred by 

```cpp
auto capture_args = std::make_shared<variant_map_type>(_arguments);
```

which will expire soon after the `run_toolkit_background` ends. So the reference value refers to an invalid memory section (the value referred by the shared pointer might be deleted or overridden).

```cpp
    return toolkit_function_response_future(
         [toolkit_name, _arguments]() {
            return this->run_toolkit(toolkit_name, _arguments);
         });
```
explicitly capture by value, even though it looks not elegant.

Reply:
How is it not safe?  The shared_ptr creates a copy of the variant_map_type, which lasts independent of the reference.